In [1]:
import pandas as pd  
import numpy as np
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')

%matplotlib inline
%config InlineBackend.figure_format = 'retina'

In [2]:
csv = '../datasets/combined/data/sentiment/num_label_tweets.csv'
my_df = pd.read_csv(csv)#,index_col=0)
my_df.head()

text        category  \
0  people will give panamaverdict in na electionw...  panama_verdict   
1  just bark dog where do panama expose corruptio...  panama_verdict   
2  It be clear from maryam nawaz bitter na campai...  panama_verdict   
3  nab references trigger by panama verdict be ha...  panama_verdict   
4  jaise motu gang saath hua tha he do not accept...  panama_verdict   

   sentiment  
0          1  
1          2  
2          1  
3          0  
4          2

In [3]:
my_df.dropna(inplace=True)
my_df.reset_index(drop=True,inplace=True)
my_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660645 entries, 0 to 660644
Data columns (total 3 columns):
text         660645 non-null object
category     660645 non-null object
sentiment    660645 non-null int64
dtypes: int64(1), object(2)
memory usage: 15.1+ MB


In [4]:
my_df.sentiment.value_counts()

2    390792
1    161068
0    108785
Name: sentiment, dtype: int64

In [7]:
my_df[my_df.category == 'cpec'].head()

text category sentiment
236846  cpec game changer for region nawaz sharif duba...     cpec  negative
236847  good question china sponsor pakistan to keep t...     cpec  positive
236848  cpececo new ecomic zone change the world econo...     cpec  positive
236849  cpec be now already success those do terrorism...     cpec   neutral
236850                                      score at cpec     cpec   neutral

In [8]:
my_df[my_df.category == 'dengue'].head()

text category sentiment
600773  just play cardboard castle dengue fever the de...   dengue   neutral
600774  big gigs for January dengue fever johnny cash ...   dengue   neutral
600775  st paul dengue fever saturday and more prior a...   dengue  positive
600776  diferencias zica dengue chicungunya pictwitter...   dengue   neutral
600777  eat more bacon no let up dengue mosquito larva...   dengue  positive

In [16]:
my_df['sentiment'] = my_df['sentiment'].map({'negative': 0, 'positive': 1, 'neutral': 2})

In [17]:
my_df.sentiment.value_counts()

2    390792
1    161068
0    108785
Name: sentiment, dtype: int64

In [18]:
my_df.head()

text        category  \
0  people will give panamaverdict in na electionw...  panama_verdict   
1  just bark dog where do panama expose corruptio...  panama_verdict   
2  It be clear from maryam nawaz bitter na campai...  panama_verdict   
3  nab references trigger by panama verdict be ha...  panama_verdict   
4  jaise motu gang saath hua tha he do not accept...  panama_verdict   

   sentiment  
0          1  
1          2  
2          1  
3          0  
4          2

In [19]:
my_df.to_csv('../datasets/combined/data/sentiment/num_label_tweets.csv',index=False)

In [20]:
csv = '../datasets/combined/data/sentiment/num_label_tweets.csv'
my_df = pd.read_csv(csv)#,index_col=0)
my_df.head()

text        category  \
0  people will give panamaverdict in na electionw...  panama_verdict   
1  just bark dog where do panama expose corruptio...  panama_verdict   
2  It be clear from maryam nawaz bitter na campai...  panama_verdict   
3  nab references trigger by panama verdict be ha...  panama_verdict   
4  jaise motu gang saath hua tha he do not accept...  panama_verdict   

   sentiment  
0          1  
1          2  
2          1  
3          0  
4          2

In [5]:
import csv
nb_classes =4
x = []
y_train = []

with open('../datasets/combined/data/sentiment/num_label_tweets.csv', 'r') as train_file:
    reader = csv.reader(train_file)
    next(reader, None)
    for row in reader:
        text = row[0]
        target = row[2]
        
        x.append(text)
        y_train.append(target)

In [6]:
print(x[10],y_train[10])

stopattackingjudiciary trump indialostpmlninvestme aassaa aa panamaverdict justicekhosa pakistan dramasharif hamdamstatusaa aac aa 2


In [7]:
import keras
#from keras import np_utils
#`x = my_df.text
y = keras.utils.to_categorical(y_train, nb_classes)

Using TensorFlow backend.


In [15]:
print(y)

<generator object <genexpr> at 0x7fede3685990>


In [8]:
from sklearn.cross_validation import train_test_split
SEED = 2000
x_train, x_validation_and_test, y_train, y_validation_and_test = train_test_split(x, y, test_size=.02, random_state=SEED)
x_validation, x_test, y_validation, y_test = train_test_split(x_validation_and_test, y_validation_and_test, test_size=.5, random_state=SEED)

/home/alizeb/.local/lib/python3.5/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


## Aritificial Neural Network

### ANN with Tfidf vectorizer

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer
tvec1 = TfidfVectorizer(max_features=100000,ngram_range=(1, 3))
tvec1.fit(x_train)

TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=100000, min_df=1,
        ngram_range=(1, 3), norm='l2', preprocessor=None, smooth_idf=True,
        stop_words=None, strip_accents=None, sublinear_tf=False,
        token_pattern='(?u)\\b\\w\\w+\\b', tokenizer=None, use_idf=True,
        vocabulary=None)

In [16]:
x_train_tfidf = tvec1.transform(x_train)

/home/alizeb/.local/lib/python3.5/site-packages/sklearn/feature_extraction/text.py:1089: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  if hasattr(X, 'dtype') and np.issubdtype(X.dtype, np.float):


In [17]:
x_validation_tfidf = tvec1.transform(x_validation).toarray()

/home/alizeb/.local/lib/python3.5/site-packages/sklearn/feature_extraction/text.py:1089: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  if hasattr(X, 'dtype') and np.issubdtype(X.dtype, np.float):


In [20]:
%%time
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression()
clf.fit(x_train_tfidf, y_train1)

NameError: name 'y_train1' is not defined

In [18]:
clf.score(x_train_tfidf, y_train)

NameError: name 'clf' is not defined

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
bvec = CountVectorizer(max_features=80000,ngram_range=(1, 2))
bvec.fit(x_train)

In [9]:
seed = 7
np.random.seed(seed)
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.layers import Flatten
from keras.layers.embeddings import Embedding
from keras.preprocessing import sequence

In [10]:
np.shape(y)[0]

660645

In [11]:
index = np.arange(np.shape(y_train)[0])
index

array([     0,      1,      2, ..., 647429, 647430, 647431])

In [12]:
index_batch = index[32*0:32*(0+1)]

In [66]:
index_batch

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31])

In [67]:
X_batch1 = x_train_tfidf[index_batch,:].toarray()
X_batch1

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [77]:
y_batch1 = y_train[index_batch]
y_batch1

array([[0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0.

In [75]:
y_train

array([[0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0., 0.],
       ...,
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 1., 0., 0., 0.]])

In [13]:
def batch_generator(X_data, y_data, batch_size):
    samples_per_epoch = X_data.shape[0]
    number_of_batches = samples_per_epoch/batch_size
    counter=0
    index = np.arange(np.shape(y_data)[0])
    while 1:
        index_batch = index[batch_size*counter:batch_size*(counter+1)]
        X_batch = X_data[index_batch,:].toarray()
        y_batch = y_data[index_batch]
        #y_batch = y_data#[y_data.index[index_batch]]
        counter += 1
        yield X_batch,y_batch
        if (counter > number_of_batches):
            counter=0

In [21]:
%%time
model = Sequential()
model.add(Dense(64, activation='relu', input_dim=100000))
model.add(Dense(nb_classes, activation='softmax'))
model.compile(optimizer='adam',
              loss= 'categorical_crossentropy',
              metrics=['accuracy'])

"""model.fit(x_train_tfidf, y_train,
    batch_size=32,
    epochs=5,
    verbose=1,
    #validation_split=0.1,
    validation_data=(x_validation_tfidf, y_validation),
         # steps_per_epoch=x_train_tfidf.shape[0]/32,
    shuffle=True)"""
model.fit_generator(generator=batch_generator(x_train_tfidf, y_train, 128),
                   epochs=5, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/128)

Epoch 1/5
5059/5058 [==============================] - 462s - loss: 0.2562 - acc: 0.9109 - val_loss: 0.1446 - val_acc: 0.9546
Epoch 2/5
5059/5058 [==============================] - 456s - loss: 0.0867 - acc: 0.9734 - val_loss: 0.1323 - val_acc: 0.9600
Epoch 3/5
5059/5058 [==============================] - 459s - loss: 0.0455 - acc: 0.9873 - val_loss: 0.1382 - val_acc: 0.9616
Epoch 4/5
5059/5058 [==============================] - 461s - loss: 0.0221 - acc: 0.9944 - val_loss: 0.1583 - val_acc: 0.9628
Epoch 5/5
5059/5058 [==============================] - 464s - loss: 0.0102 - acc: 0.9976 - val_loss: 0.1818 - val_acc: 0.9600
CPU times: user 22min 53s, sys: 17min 51s, total: 40min 44s
Wall time: 38min 26s


In [22]:
print(model.summary())
from keras.models import model_from_json
# serialize model to JSON
model_json = model.to_json()
with open("../datasets/combined/models/sentiment/keras_model.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model.save_weights("../datasets/combined/models/sentiment/keras_model.h5")
print("Saved model to disk")

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 64)                6400064   
_________________________________________________________________
dense_4 (Dense)              (None, 4)                 260       
Total params: 6,400,324
Trainable params: 6,400,324
Non-trainable params: 0
_________________________________________________________________
None
Saved model to disk


In [ ]:
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/sentiment/keras_model.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
loaded_model = model_from_json(loaded_model_json)
# load weights into new model
loaded_model.load_weights("../datasets/combined/models/sentiment/keras_model.h5")
print("Loaded model from disk")

It looks like the model had the best validation accuracy after 2 epochs, and after that, it fails to generalize so validation accuracy slowly decreases, while training accuracy increases. But if you remember the result I got from logistic regression (train accuracy: 84.19%, validation accuracy: 82.91%), you can see that the above neural network failed to outperform logistic regression in terms of validation.

Let's see if normalizing inputs have any effect on the performance.

In [23]:
from sklearn.preprocessing import Normalizer
norm = Normalizer().fit(x_train_tfidf)
x_train_tfidf_norm = norm.transform(x_train_tfidf)
x_validation_tfidf_norm = norm.transform(x_validation_tfidf)

In [26]:
%%time
model_n = Sequential()
model_n.add(Dense(64, activation='relu', input_dim=100000))
model_n.add(Dense(nb_classes, activation='sigmoid'))
model_n.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model_n.fit_generator(generator=batch_generator(x_train_tfidf_norm, y_train, 32),
                    epochs=5, validation_data=(x_validation_tfidf_norm, y_validation),
                    steps_per_epoch=x_train_tfidf_norm.shape[0]/32)

Epoch 1/5
20233/20232 [==============================] - 623s - loss: 0.1120 - acc: 0.9582 - val_loss: 0.0685 - val_acc: 0.9768
Epoch 2/5
20233/20232 [==============================] - 624s - loss: 0.0462 - acc: 0.9848 - val_loss: 0.0638 - val_acc: 0.9799
Epoch 3/5
20233/20232 [==============================] - 601s - loss: 0.0297 - acc: 0.9908 - val_loss: 0.0672 - val_acc: 0.9796
Epoch 4/5
20233/20232 [==============================] - 601s - loss: 0.0204 - acc: 0.9941 - val_loss: 0.0728 - val_acc: 0.9804
Epoch 5/5
20233/20232 [==============================] - 601s - loss: 0.0140 - acc: 0.9961 - val_loss: 0.0804 - val_acc: 0.9796
CPU times: user 58min 22s, sys: 10min 43s, total: 1h 9min 6s
Wall time: 50min 53s


In [27]:
from keras.models import model_from_json
# serialize model to JSON
model_json = model_n.to_json()
with open("../datasets/combined/models/sentiment/keras_model_n.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model.save_weights("../datasets/combined/models/sentiment/keras_model_n.h5")
print("Saved model to disk")

Saved model to disk


In [ ]:
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/sentiment/keras_model_n.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model_n = model_from_json(loaded_model_json)
# load weights into new model
model_n.load_weights("../datasets/combined/models/sentiment/keras_model_n.h5")
print("Loaded model from disk")

In [16]:
%%time
model1 = Sequential()

model1.add(Dense(64, activation='relu', input_dim=100000))
model1.add(Dropout(0.2))
model1.add(Dense(nb_classes, activation='sigmoid'))
model1.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model1.fit_generator(generator=batch_generator(x_train_tfidf, y_train, 32),
                    epochs=5, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/32)

Epoch 1/5
29561/29560 [==============================] - 922s - loss: 0.0105 - acc: 0.9968 - val_loss: 0.0027 - val_acc: 0.9987
Epoch 2/5
29561/29560 [==============================] - 927s - loss: 0.0022 - acc: 0.9991 - val_loss: 0.0032 - val_acc: 0.9987
Epoch 3/5
29561/29560 [==============================] - 928s - loss: 0.0019 - acc: 0.9992 - val_loss: 0.0035 - val_acc: 0.9987
Epoch 4/5
29561/29560 [==============================] - 932s - loss: 0.0017 - acc: 0.9992 - val_loss: 0.0038 - val_acc: 0.9985
Epoch 5/5
29561/29560 [==============================] - 930s - loss: 0.0017 - acc: 0.9993 - val_loss: 0.0041 - val_acc: 0.9985
CPU times: user 1h 26min 47s, sys: 15min 55s, total: 1h 42min 42s
Wall time: 1h 17min 22s


In [18]:
print(model1.summary())
# serialize model to JSON
model_json = model1.to_json()
with open("../datasets/combined/models/sentiment/keras_model1.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model1.save_weights("../datasets/combined/models/sentiment/keras_model1.h5")
print("Saved model to disk")

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 64)                6400064   
_________________________________________________________________
dropout_1 (Dropout)          (None, 64)                0         
_________________________________________________________________
dense_4 (Dense)              (None, 6)                 390       
Total params: 6,400,454
Trainable params: 6,400,454
Non-trainable params: 0
_________________________________________________________________
None
Saved model to disk


In [ ]:
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/sentiment/keras_model1.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model1 = model_from_json(loaded_model_json)
# load weights into new model
model1.load_weights("../datasets/combined/models/sentiment/keras_model1.h5")
print("Loaded model from disk")

In [11]:
def batch_generator_shuffle(X_data, y_data, batch_size):
    samples_per_epoch = X_data.shape[0]
    number_of_batches = samples_per_epoch/batch_size
    counter=0
    index = np.arange(np.shape(y_data)[0])
    np.random.shuffle(index)
    while 1:
        index_batch = index[batch_size*counter:batch_size*(counter+1)]
        X_batch = X_data[index_batch,:].toarray()
        y_batch = y_data[index_batch]#[y_data.index[index_batch]]
        counter += 1
        yield X_batch,y_batch
        if (counter > number_of_batches):
            np.random.shuffle(index)
            counter=0

In [21]:
%%time
model_s = Sequential()
model_s.add(Dense(64, activation='relu', input_dim=100000))
model_s.add(Dense(nb_classes, activation='sigmoid'))
model_s.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model_s.fit_generator(generator=batch_generator_shuffle(x_train_tfidf, y_train, 32),
                    epochs=5, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/32)

Epoch 1/5
29561/29560 [==============================] - 910s - loss: 0.0097 - acc: 0.9970 - val_loss: 0.0031 - val_acc: 0.9987
Epoch 2/5
29561/29560 [==============================] - 915s - loss: 0.0019 - acc: 0.9991 - val_loss: 0.0032 - val_acc: 0.9986
Epoch 3/5
29561/29560 [==============================] - 908s - loss: 0.0017 - acc: 0.9992 - val_loss: 0.0035 - val_acc: 0.9987
Epoch 4/5
29561/29560 [==============================] - 912s - loss: 0.0015 - acc: 0.9992 - val_loss: 0.0040 - val_acc: 0.9985
Epoch 5/5
29561/29560 [==============================] - 897s - loss: 0.0014 - acc: 0.9993 - val_loss: 0.0041 - val_acc: 0.9985
CPU times: user 1h 25min 7s, sys: 15min 41s, total: 1h 40min 49s
Wall time: 1h 15min 44s


In [22]:
# serialize model to JSON
model_json = model_s.to_json()
with open("../datasets/combined/models/sentiment/keras_model_s.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model_s.save_weights("../datasets/combined/models/sentiment/keras_model_s.h5")
print("Saved model to disk")

Saved model to disk


In [ ]:
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/sentiment/keras_model_s.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model_s = model_from_json(loaded_model_json)
# load weights into new model
model_s.load_weights("../datasets/combined/models/sentiment/keras_model_s.h5")
print("Loaded model from disk")

In [23]:
%%time
model_s_1 = Sequential()
model_s_1.add(Dense(64, activation='relu', input_dim=100000))
model_s_1.add(Dropout(0.2))
model_s_1.add(Dense(nb_classes, activation='sigmoid'))
model_s_1.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model_s_1.fit_generator(generator=batch_generator_shuffle(x_train_tfidf, y_train, 32),
                    epochs=5, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/32)

Epoch 1/5
29561/29560 [==============================] - 925s - loss: 0.0107 - acc: 0.9967 - val_loss: 0.0030 - val_acc: 0.9987
Epoch 2/5
29561/29560 [==============================] - 968s - loss: 0.0021 - acc: 0.9991 - val_loss: 0.0034 - val_acc: 0.9989
Epoch 3/5
29561/29560 [==============================] - 901s - loss: 0.0019 - acc: 0.9991 - val_loss: 0.0035 - val_acc: 0.9988
Epoch 4/5
29561/29560 [==============================] - 915s - loss: 0.0018 - acc: 0.9992 - val_loss: 0.0042 - val_acc: 0.9985
Epoch 5/5
29561/29560 [==============================] - 914s - loss: 0.0017 - acc: 0.9992 - val_loss: 0.0041 - val_acc: 0.9986
CPU times: user 1h 27min 32s, sys: 15min 41s, total: 1h 43min 14s
Wall time: 1h 17min 6s


In [24]:
# serialize model to JSON
model_json = model_s_1.to_json()
with open("../datasets/combined/models/classification/keras_model_s_1.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model_s_1.save_weights("../datasets/combined/models/classification/keras_model_s_1.h5")
print("Saved model to disk")

Saved model to disk


In [ ]:
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/classification/keras_model_s_1.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model_s_1 = model_from_json(loaded_model_json)
# load weights into new model
model_s_1.load_weights("../datasets/combined/models/classification/keras_model_s_1.h5")
print("Loaded model from disk")

In [25]:
%%time
import keras
custom_adam = keras.optimizers.Adam(lr=0.005, beta_1=0.9, beta_2=0.999, epsilon=1e-8)
model_testing_2 = Sequential()
model_testing_2.add(Dense(64, activation='relu', input_dim=100000))
model_testing_2.add(Dense(nb_classes, activation='sigmoid'))
model_testing_2.compile(optimizer=custom_adam,
              loss='binary_crossentropy',
              metrics=['accuracy'])

model_testing_2.fit_generator(generator=batch_generator_shuffle(x_train_tfidf, y_train, 32),
                    epochs=2, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/32)

Epoch 1/2
29561/29560 [==============================] - 901s - loss: 0.0050 - acc: 0.9982 - val_loss: 0.0028 - val_acc: 0.9988
Epoch 2/2
29561/29560 [==============================] - 902s - loss: 0.0021 - acc: 0.9991 - val_loss: 0.0037 - val_acc: 0.9986
CPU times: user 33min 50s, sys: 6min 10s, total: 40min
Wall time: 30min 4s


In [26]:
# serialize model to JSON
model_json = model_testing_2.to_json()
with open("../datasets/combined/models/classification/keras_model_testing_2.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model_testing_2.save_weights("../datasets/combined/models/classification/keras_model_testing_2.h5")
print("Saved model to disk")

Saved model to disk


In [ ]:
 
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/classification/keras_model_testing_2.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model_testing_2 = model_from_json(loaded_model_json)
# load weights into new model
model_testing_2.load_weights("../datasets/combined/models/classification/keras_model_testing_2.h5")
print("Loaded model from disk")

In [27]:
%%time
custom_adam = keras.optimizers.Adam(lr=0.01, beta_1=0.9, beta_2=0.999, epsilon=1e-8)
model_testing_3 = Sequential()
model_testing_3.add(Dense(64, activation='relu', input_dim=100000))
model_testing_3.add(Dense(nb_classes, activation='sigmoid'))
model_testing_3.compile(optimizer=custom_adam,
              loss='binary_crossentropy',
              metrics=['accuracy'])

model_testing_3.fit_generator(generator=batch_generator_shuffle(x_train_tfidf, y_train, 32),
                    epochs=2, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/32)

Epoch 1/2
29561/29560 [==============================] - 917s - loss: 0.0048 - acc: 0.9983 - val_loss: 0.0036 - val_acc: 0.9985
Epoch 2/2
29561/29560 [==============================] - 918s - loss: 0.0023 - acc: 0.9991 - val_loss: 0.0039 - val_acc: 0.9987
CPU times: user 34min 21s, sys: 6min 14s, total: 40min 36s
Wall time: 30min 36s


In [28]:
# serialize model to JSON
model_json = model_testing_3.to_json()
with open("../datasets/combined/models/classification/keras_model_testing_3.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model_testing_3.save_weights("../datasets/combined/models/classification/keras_model_testing_3.h5")
print("Saved model to disk")

Saved model to disk


In [ ]:
 
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/classification/keras_model_testing_3.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model_testing_3 = model_from_json(loaded_model_json)
# load weights into new model
model_testing_3.load_weights("../datasets/combined/models/classification/keras_model_testing_3.h5")
print("Loaded model from disk")

In [29]:
%%time
custom_adam = keras.optimizers.Adam(lr=0.1, beta_1=0.9, beta_2=0.999, epsilon=1e-8)
model_testing_4 = Sequential()
model_testing_4.add(Dense(64, activation='relu', input_dim=100000))
model_testing_4.add(Dense(nb_classes, activation='sigmoid'))
model_testing_4.compile(optimizer=custom_adam,
              loss='binary_crossentropy',
              metrics=['accuracy'])

model_testing_4.fit_generator(generator=batch_generator_shuffle(x_train_tfidf, y_train, 32),
                    epochs=2, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/32)

Epoch 1/2
29561/29560 [==============================] - 910s - loss: 0.0084 - acc: 0.9979 - val_loss: 0.0069 - val_acc: 0.9971
Epoch 2/2
29561/29560 [==============================] - 915s - loss: 0.0051 - acc: 0.9986 - val_loss: 0.0058 - val_acc: 0.9980
CPU times: user 34min 17s, sys: 6min 11s, total: 40min 29s
Wall time: 30min 27s


In [32]:
print(model_testing_4.summary())
# serialize model to JSON
model_json = model_testing_4.to_json()
with open("../datasets/combined/models/classification/keras_model_testing_4.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model_testing_4.save_weights("../datasets/combined/models/classification/keras_model_testing_4.h5")
print("Saved model to disk")

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 64)                6400064   
_________________________________________________________________
dense_16 (Dense)             (None, 6)                 390       
Total params: 6,400,454
Trainable params: 6,400,454
Non-trainable params: 0
_________________________________________________________________
None
Saved model to disk


In [ ]:
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/classification/keras_model_testing_4.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model_testing_4 = model_from_json(loaded_model_json)
# load weights into new model
model_testing_4.load_weights("../datasets/combined/models/classification/keras_model_testing_4.h5")
print("Loaded model from disk")

In [33]:
%%time
custom_adam = keras.optimizers.Adam(lr=0.0005, beta_1=0.9, beta_2=0.999, epsilon=1e-8)
model_testing_5 = Sequential()
model_testing_5.add(Dense(64, activation='relu', input_dim=100000))
model_testing_5.add(Dense(nb_classes, activation='sigmoid'))
model_testing_5.compile(optimizer=custom_adam,
              loss='binary_crossentropy',
              metrics=['accuracy'])

model_testing_5.fit_generator(generator=batch_generator_shuffle(x_train_tfidf, y_train, 32),
                    epochs=5, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/32)

Epoch 1/5
29561/29560 [==============================] - 915s - loss: 0.0149 - acc: 0.9955 - val_loss: 0.0034 - val_acc: 0.9986
Epoch 2/5
29561/29560 [==============================] - 917s - loss: 0.0022 - acc: 0.9991 - val_loss: 0.0033 - val_acc: 0.9986
Epoch 3/5
29561/29560 [==============================] - 918s - loss: 0.0018 - acc: 0.9992 - val_loss: 0.0034 - val_acc: 0.9985
Epoch 4/5
29561/29560 [==============================] - 929s - loss: 0.0016 - acc: 0.9992 - val_loss: 0.0035 - val_acc: 0.9985
Epoch 5/5
29561/29560 [==============================] - 925s - loss: 0.0015 - acc: 0.9992 - val_loss: 0.0038 - val_acc: 0.9984
CPU times: user 1h 26min 38s, sys: 15min 8s, total: 1h 41min 47s
Wall time: 1h 16min 46s


In [34]:
print(model_testing_5.summary())
# serialize model to JSON
model_json = model_testing_5.to_json()
with open("../datasets/combined/models/classification/keras_model_testing_5.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model_testing_5.save_weights("../datasets/combined/models/classification/keras_model_testing_5.h5")
print("Saved model to disk")

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_17 (Dense)             (None, 64)                6400064   
_________________________________________________________________
dense_18 (Dense)             (None, 6)                 390       
Total params: 6,400,454
Trainable params: 6,400,454
Non-trainable params: 0
_________________________________________________________________
None
Saved model to disk


In [ ]:
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/classification/keras_model_testing_5.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model_testing_5 = model_from_json(loaded_model_json)
# load weights into new model
model_testing_5.load_weights("../datasets/combined/models/classification/keras_model_testing_5.h5")
print("Loaded model from disk")

In [12]:
%%time
model_s_2 = Sequential()
model_s_2.add(Dense(128, activation='relu', input_dim=100000))
model_s_2.add(Dense(nb_classes, activation='sigmoid'))
model_s_2.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model_s_2.fit_generator(generator=batch_generator_shuffle(x_train_tfidf, y_train, 32),
                    epochs=2, validation_data=(x_validation_tfidf, y_validation),
                    steps_per_epoch=x_train_tfidf.shape[0]/32)

Epoch 1/2
29561/29560 [==============================] - 1340s - loss: 0.0079 - acc: 0.9975 - val_loss: 0.0030 - val_acc: 0.9988
Epoch 2/2
29561/29560 [==============================] - 1313s - loss: 0.0019 - acc: 0.9991 - val_loss: 0.0031 - val_acc: 0.9987
CPU times: user 41min 13s, sys: 11min 24s, total: 52min 37s
Wall time: 44min 14s


In [13]:
print(model_s_2.summary())
from keras.models import model_from_json
# serialize model to JSON
model_json = model_s_2.to_json()
with open("../datasets/combined/models/classification/keras_model_s_2.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model_s_2.save_weights("../datasets/combined/models/classification/keras_model_s_2.h5")
print("Saved model to disk")

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 128)               12800128  
_________________________________________________________________
dense_2 (Dense)              (None, 6)                 774       
Total params: 12,800,902
Trainable params: 12,800,902
Non-trainable params: 0
_________________________________________________________________
None
Saved model to disk


In [ ]:
# later...
 
# load json and create model
json_file = open('../datasets/combined/models/classification/keras_model_s_2.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
model_s_2 = model_from_json(loaded_model_json)
# load weights into new model
model_s_2.load_weights("../datasets/combined/models/classification/keras_model_s_2.h5")
print("Loaded model from disk")

| model | learning rate | input layer (nodes) | data shuffling | hidden layer (nodes) | dropout | output layer (nodes) | training accuracy | validation accuracy |
|-----|------|---|-----|----|----|----|----|
| ANN_1 | 0.001 | 1 (100,000)  | X | 1 (64) relu  |  X  | 1 (1) softmax   | 99.82% | 99.53% |
| ANN_2 | 0.001 | 1 (100,000)  | X | 1 (64) relu  |  0.2  | 1 (1) sigmoid   | 99.93% | 99.85% |
| ANN_3 | 0.001 | 1 (100,000)  | O | 1 (64) relu  |  X  | 1 (1) sigmoid   | 99.93% | 99.85% |
| ANN_4 | 0.001 | 1 (100,000)  | O | 1 (64) relu  |  0.2  | 1 (1) sigmoid   | 99.92% | 99.86%  |
| ANN_5 | 0.0005 | 1 (100,000)  | O | 1 (64) relu  |  X  | 1 (1) sigmoid   | 99.91% | 99.87%  |
| ANN_6 | 0.005 | 1 (100,000)  | O | 1 (64) relu  |  X  | 1 (1) sigmoid   | 99.86% | 99.87%  |
| ANN_7 | 0.01 | 1 (100,000)  | O | 1 (64) relu  |  X  | 1 (1) sigmoid   | 99.86% | 99.80%  |
| ANN_8 | 0.1 | 1 (100,000)  | O | 1 (64) relu  |  X  | 1 (1) sigmoid   | 99.92% | 99.84%  |
| ANN_9 | 0.001 | 1 (100,000)  | O | 1 (128) relu  |  X  | 1 (1) sigmoid   | 99.91% | 99.87%  |